lldb/source/Plugins/TypeSystem/Swift/SwiftASTContext.cpp

```diff
@@ -5315,7 +5315,7 @@ ConstString SwiftASTContext::GetTypeName(opaque_compiler_type_t type) {
         swift_type.transform([](swift::Type type) -> swift::Type {
           if (swift::SyntaxSugarType *syntax_sugar_type =
                   swift::dyn_cast<swift::SyntaxSugarType>(type.getPointer())) {
-            return syntax_sugar_type->getSinglyDesugaredType();
+            return syntax_sugar_type->getDesugaredType();
```

Comment:
This looks good! I think we will also need to remove the workaround from the 3 switch-cases at

https://github.com/apple/llvm-project/blob/a5de9f53c1725aff7c70868c59a0e76ae011af70/lldb/source/Plugins/TypeSystem/Swift/TypeSystemSwiftTypeRef.cpp#L581

otherwise this should trigger a validation error when comparing the output of TypeSystemSwiftTypeRef and SwiftASTContext, right?
The fact that it didn't trigger one for you makes me wonder if we are missing a testcase with a nested Optional in the test suite...

Reply:
which tests are you referring to?

Reply:
I meant that this shows that (perhaps this indicates that) not a single test in lldb/test/API/lang/swift exercises this code path since this patch didn't trigger a validation error.

Reply:
I'll look for where and how to cover this with tests.

Reply:
I spent some time on this, and it looks like `getDesugaredType` is … broken?

1. `getDesugaredType` only applies to `SugarType` instances, otherwise it returns the given type (see [include/swift/AST/Types.h#6065-L6068](https://github.com/apple/swift/blob/a7e8dbf4696370b16ca7cc9531921fabfd921b32/include/swift/AST/Types.h#L6065-L6068))
2. for sugared types, `getDesugaredType` calls `getSinglyDesugaredType()->getDesugaredType()`, which seems like its recursive
3. `getSinglyDesugaredType` returns a `BoundGenericType` **not a `SugarType`** (see [lib/AST/Type.cpp#L1422-L1426](https://github.com/apple/swift/blob/a7e8dbf4696370b16ca7cc9531921fabfd921b32/lib/AST/Type.cpp#L1422-L1426))

It seems the recursion will always fail, calling `getDesugaredType` on a `BoundGenericType` will return the same BGT.

The docstring for `getDesugaredType` seems to say it will recursively desugar:

```swift
/// getDesugaredType - If this type is a sugared type, remove all levels of
/// sugar until we get down to a non-sugar type.
TypeBase *getDesugaredType();
```

Reply:
It looks like this is meant to peel off only the outer layers of sugar and not to recursively lower all sugared types contained in the type.
Regarding (1) that seems to be the right thing, for example `T!` is a unary sugared type that de-sugars to `Swift.Optional<T>` (a bound generic type).

Now that I'm thinking about it — maybe we also don't need to remove more than the outermost layer of sugar. If you drill down into a member/child of a sugared type we would do that by creating a new ValueObject with a type specifically for the member/child. There again any data formatters would only match the outermost type. Or — we could instead call `Type.subst()` and replace each child type with its singly-desugared type.

What do you think?

Reply:
At this point I'd back up a step and first ask: why desugar?

Reply:
My understanding is that the main reason is that the data formatters match on the type name and we have data formatters for Swift.Optional, Swift.Dictionary, (and possibly others). The secondary reason would be to canonicalize types before comparing for equivalence / matching / caching, ... . `GetTypeName()` probably _doesn't_ want to canonicalize though since we probably care about the difference between, e.g., `typealias bar = foo` and `foo` where we call `GetTypeName()`.

Reply:
thanks, data formatters definitely makes sense.

> The secondary reason would be to canonicalize types before comparing for equivalence / matching / caching

in theory, wouldn't these be better done with full fidelity values and types, instead of a string representation